File: app/src/main/java/com/hana053/micropost/ui/NavigatorImpl.java
package com.hana053.micropost.ui;

import android.app.Activity;
import android.content.Intent;

import com.hana053.micropost.domain.User;
import com.hana053.micropost.ui.pages.login.LoginActivity;
import com.hana053.micropost.ui.pages.main.MainActivity;
import com.hana053.micropost.ui.pages.relateduserlist.followerlist.FollowerListActivity;
import com.hana053.micropost.ui.pages.relateduserlist.followinglist.FollowingListActivity;
import com.hana053.micropost.ui.pages.signup.SignupActivity;
import com.hana053.micropost.ui.pages.usershow.UserShowActivity;

import org.parceler.Parcels;

public class NavigatorImpl implements Navigator {

    private final Activity activity;

    public NavigatorImpl(Activity activity) {
        this.activity = activity;
    }

    @Override
    public void navigateToUserShow(User user) {
        final Intent intent = new Intent(activity, UserShowActivity.class);
        intent.putExtra(UserShowActivity.KEY_USER, Parcels.wrap(user));
        activity.startActivity(intent);
    }

    @Override
    public void navigateToMain() {
        final Intent intent = new Intent(activity, MainActivity.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
        activity.startActivity(intent);
    }

    @Override
    public void navigateToSignup() {
        final Intent intent = new Intent(activity, SignupActivity.class);
        activity.startActivity(intent);
    }

    @Override
    public void navigateToLogin() {
        final Intent intent = new Intent(activity, LoginActivity.class);
        activity.startActivity(intent);
    }

    @Override
    public void navigateToFollowings(long userId) {
        final Intent intent = new Intent(activity, FollowingListActivity.class);
        intent.putExtra(FollowingListActivity.KEY_USER_ID, userId);
        activity.startActivity(intent);
    }

    @Override
    public void navigateToFollowers(long userId) {
        final Intent intent = new Intent(activity, FollowerListActivity.class);
        intent.putExtra(FollowerListActivity.KEY_USER_ID, userId);
        activity.startActivity(intent);
    }
}

File: app/src/main/java/com/hana053/micropost/activity/NavigatorImpl.kt
package com.hana053.micropost.activity

import android.app.Activity
import android.content.Intent
import com.hana053.micropost.R
import com.hana053.micropost.pages.login.LoginActivity
import com.hana053.micropost.pages.main.MainActivity
import com.hana053.micropost.pages.main.MainActivity.Companion.REQUEST_POST
import com.hana053.micropost.pages.micropostnew.MicropostNewActivity
import com.hana053.micropost.pages.relateduserlist.RelatedUserListActivity
import com.hana053.micropost.pages.relateduserlist.RelatedUserListActivity.ListType.FOLLOWER
import com.hana053.micropost.pages.relateduserlist.RelatedUserListActivity.ListType.FOLLOWING
import com.hana053.micropost.pages.signup.SignupActivity
import com.hana053.micropost.pages.top.TopActivity
import com.hana053.micropost.pages.usershow.UserShowActivity
import java.lang.ref.WeakReference

class NavigatorImpl(activity: Activity) : Navigator {

    private val activity = WeakReference<Activity>(activity)

    override fun navigateToTop() {
        val intent = Intent(activity.get(), TopActivity::class.java)
        intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_NEW_TASK
        activity.get()?.startActivity(intent)
    }

    override fun navigateToMain() {
        val intent = Intent(activity.get(), MainActivity::class.java)
        intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
        activity.get()?.startActivity(intent)
    }

    override fun navigateToLogin() {
        val intent = Intent(activity.get(), LoginActivity::class.java)
        activity.get()?.startActivity(intent)
    }

    override fun navigateToSignup() {
        val intent = Intent(activity.get(), SignupActivity::class.java)
        activity.get()?.startActivity(intent)
    }

    override fun navigateToUserShow(userId: Long) {
        val intent = Intent(activity.get(), UserShowActivity::class.java)
        intent.putExtra(UserShowActivity.KEY_USER_ID, userId)
        activity.get()?.startActivity(intent)
    }

    override fun navigateToFollowerList(userId: Long) {
        val intent = Intent(activity.get(), RelatedUserListActivity::class.java)
        intent.putExtra(RelatedUserListActivity.KEY_USER_ID, userId)
        intent.putExtra(RelatedUserListActivity.KEY_LIST_TYPE, FOLLOWER)
        activity.get()?.startActivity(intent)
    }

    override fun navigateToFollowingList(userId: Long) {
        val intent = Intent(activity.get(), RelatedUserListActivity::class.java)
        intent.putExtra(RelatedUserListActivity.KEY_USER_ID, userId)
        intent.putExtra(RelatedUserListActivity.KEY_LIST_TYPE, FOLLOWING)
        activity.get()?.startActivity(intent)
    }

    override fun navigateToMicropostNew() {
        val intent = Intent(activity.get(), MicropostNewActivity::class.java)
        activity.get()?.startActivityForResult(intent, REQUEST_POST)
        activity.get()?.overridePendingTransition(R.anim.slide_in_up, 0)
    }
}
